include/mockturtle/networks/cover.hpp

```diff
@@ -0,0 +1,905 @@
+    detail::foreach_element_transform<IteratorType, uint32_t>(
+        _storage->outputs.begin() + _storage->data.num_pos, _storage->outputs.end(), []( auto o ) { return o.index; },
+        fn );
+  }
+
+  template<typename Fn>
+  void foreach_register( Fn&& fn ) const
+  {
+    static_assert( detail::is_callable_with_index_v<Fn, std::pair<signal, node>, void> ||
+                   detail::is_callable_without_index_v<Fn, std::pair<signal, node>, void> ||
+                   detail::is_callable_with_index_v<Fn, std::pair<signal, node>, bool> ||
+                   detail::is_callable_without_index_v<Fn, std::pair<signal, node>, bool> );
+
+    assert( _storage->inputs.size() - _storage->data.num_pis == _storage->outputs.size() - _storage->data.num_pos );
+    auto ro = _storage->inputs.begin() + _storage->data.num_pis;
+    auto ri = _storage->outputs.begin() + _storage->data.num_pos;
+    if constexpr ( detail::is_callable_without_index_v<Fn, std::pair<signal, node>, bool> )
+    {
+      while ( ro != _storage->inputs.end() && ri != _storage->outputs.end() )
+      {
+        if ( !fn( std::make_pair( ( ri++ )->index, ro++ ) ) )
+          return;
+      }
+    }
+    else if constexpr ( detail::is_callable_with_index_v<Fn, std::pair<signal, node>, bool> )
+    {
+      uint32_t index{ 0 };
+      while ( ro != _storage->inputs.end() && ri != _storage->outputs.end() )
+      {
+        if ( !fn( std::make_pair( ( ri++ )->index, ro++ ), index++ ) )
+          return;
+      }
+    }
+    else if constexpr ( detail::is_callable_without_index_v<Fn, std::pair<signal, node>, void> )
+    {
+      while ( ro != _storage->inputs.end() && ri != _storage->outputs.end() )
+      {
+        fn( std::make_pair( ( ri++ )->index, *ro++ ) );
+      }
+    }
+    else if constexpr ( detail::is_callable_with_index_v<Fn, std::pair<signal, node>, void> )
+    {
+      uint32_t index{ 0 };
+      while ( ro != _storage->inputs.end() && ri != _storage->outputs.end() )
+      {
+        fn( std::make_pair( ( ri++ )->index, *ro++ ), index++ );
+      }
+    }
+  }
+
+  template<typename Fn>
+  void foreach_gate( Fn&& fn ) const
+  {
+    auto r = range<uint64_t>( 2u, _storage->nodes.size() ); /* start from 2 to avoid constants */
+    detail::foreach_element_if(
+        r.begin(), r.end(),
+        [this]( auto n ) { return !is_ci( n ); },
+        fn );
+  }
+
+  template<typename Fn>
+  void foreach_fanin( node const& n, Fn&& fn ) const
+  {
+    if ( n == 0 || is_ci( n ) )
+      return;
+
+    using IteratorType = decltype( _storage->outputs.begin() );
+    detail::foreach_element_transform<IteratorType, uint32_t>(
+        _storage->nodes[n].children.begin(), _storage->nodes[n].children.end(), []( auto f ) { return f.index; },
+        fn );
+  }
+
+#pragma endregion
+
+#pragma region Simulate values
+  std::vector<bool> compute_on_node( node const& n, std::vector<kitty::cube> const& domain_cubes ) const
```

Comment:
I didn't think of it as a function to be used by the user at the time. I wanted a function to use for testing the correctness of the cover stored in a single node. In fact, the function requires as input the cubes at which one wants to know if the node evaluates to 1 or 0. I introduced this code mainly for testing purposes. I was considering the possibility to extend the implementation and to write the corresponding function for any node given the input of the network. I didn't put too much importance on this from a user perspective because we have always looked at this data structure as a "bridging object". I have always considered its utility to be the one of bypassing the computational problem you identified in the .blif reading into klut networks. As such, I thought nobody could be interested in using a cover network for direct simulation since the cover network is meant to easier the mapping to other data structures, each of which has properly working simulation functions. Let me know if:

1. I can let things as they are
2. The simulate on node could be used by somebody -> I'll write the documentation;
3. The simulate on node could be used by somebody and also a more general simulation function is needed -> I'll write the function and the documentation.

Reply:
I see this issue a bit differently. 
According to the [mockturtle philosophy](https://mockturtle.readthedocs.io/en/latest/philosophy.html), the network interface is the fundament of the library (recall that mockturtle is a "logic network library"). This means all network types can potentially be used by any algorithm -- if the required methods are provided by the network type. 
Even though we said that the motivation of implementing this network type is to bridge from large LUTs to other decomposed networks, this does not mean that it cannot be used for other purposes in the future. It is fine not to provide all of the interfaces, so that it does not fit into most algorithms now. But it is still not an arbitrary data structure and should not have arbitrary interfaces. It is also fine to provide more interfaces that the other network types do not have, but they should be well-motivated and documented.
If this function is only for testing purposes, I would put it in the testing code.

Reply:
the new version will have the _compute_ interface working as in the other network types. This substitutes _compute_on_node_.